### PR TITLE
Scheduling and Assembly Updates

### PR DESCRIPTION
**Pull request by Checkmate50**
Creating this as a pull request just for reference, if I could get a quick review (or just make sure it works before I merge).  `cargo test` in `caiman-test` works on my machine, and the frontend seems to work (note that it uses the filename extension to determine to use the assembly, specifically the `.cair` extension).

### CODE REVIEW

File: src/type_system/scheduling.rs [hunk 1]
```diff
@@ -282,7 +282,7 @@ impl<'program> FuncletChecker<'program>
 		}
 	}
 
-	fn try_transition_slot(&mut self, slot_node_id : ir::NodeId, place : ir::Place, from_to_stage_pairs : &[(ir::ResourceQueueStage, ir::ResourceQueueStage)]) -> bool
+	fn 	try_transition_slot(&mut self, slot_node_id : ir::NodeId, place : ir::Place, from_to_stage_pairs : &[(ir::ResourceQueueStage, ir::ResourceQueueStage)]) -> bool
```
Comment (od0x0) [inline src/type_system/scheduling.rs:285]: Tab snuck in

File: src/rust_wgpu_backend/codegen.rs [hunk 2]
```diff
@@ -538,7 +538,13 @@ impl<'program> CodeGen<'program>
 
 				use std::iter::FromIterator;
 
-				let argument_var_ids = Vec::from_iter(arguments.iter().enumerate().map(|(index, x)| { let slot_id = input_slot_ids[index]; placement_state.slot_variable_ids[& slot_id] })).into_boxed_slice();
+				dbg!(function);
```
Comment (od0x0) [inline src/rust_wgpu_backend/codegen.rs:541]: I think a few surgical devices are being left in the patient here

File: src/rust_wgpu_backend/code_generator.rs [hunk 3]
```diff
@@ -660,6 +660,7 @@ impl<'program> CodeGenerator<'program>
 	fn write_states(&mut self)
 	{
 		//self.state_code_writer
+		// todo: chec if removing "use caiman_rt::wgpu;" is going to break things
```
Comment (od0x0) [inline src/rust_wgpu_backend/code_generator.rs:663]: Does it?

File: src/main.rs [hunk 4]
```diff
@@ -24,18 +41,12 @@ fn compile(input_file : &mut File, output_file : &mut File, options_opt : Option
...
-			write!(output_file, "{}", output_string);
-		}
-	}
+	let result = caiman::frontend::compile_caiman
+		(& input_string, options_opt, use_assembly);
+	output_result(result, output_file);
 }
 
-fn explicate(input_file : &mut File, output_file : &mut File)
+fn explicate(input_file : &mut File, output_file : &mut Option<File>, use_assembly : bool)
```
Comment (od0x0) [inline src/main.rs:49]: Can this boolean be an enum instead? We'll almost certainly want different input types eventually and it's easy to forget what true/false means when actually using these functions.

It's ok to defer this if merging is urgent.

File: src/ir.rs [hunk 5]
```diff
@@ -335,10 +370,28 @@ pub struct Program
 	//pub shader_modules : HashMap<usize, ShaderModule>
 }
 
+impl serde::Serialize for Program {
```
Comment (od0x0) [inline src/ir.rs:373]: I don't think we need this now with Mateo's improved StableVec type

File: src/ir.rs [hunk 6]
```diff
@@ -198,7 +200,7 @@ pub enum FuncletKind
 	Value,
 	ScheduleExplicit,
 	Inline, // Adopts the constraints of the calling funclet
-	Timeline,
+Timeline,
```
Comment (od0x0) [inline src/ir.rs:203]: Not sure what happened here with the indentation, but it happened a couple times in the file and doesn't seem intentional

File: src/frontend.rs [hunk 7]
```diff
@@ -31,16 +31,31 @@ impl std::fmt::Display for CompileError
 	}
 }
 
-pub fn compile_ron_definition(input_string : &str, options_opt : Option<CompileOptions>) -> Result<String, CompileError>
+fn read_definition(input_string : &str, assembly : bool) -> Result<Definition, CompileError> {
```
Comment (od0x0) [inline src/frontend.rs:34]: This function inherits my complaints about a boolean for specifying which language to use

File: src/arena.rs [hunk 8]
```diff
@@ -0,0 +1,161 @@
+use std::collections::HashMap;
```
Comment (od0x0) [inline src/arena.rs:1]: Definitely see if things still work with Mateo's change

**Review by od0x0 — CHANGES_REQUESTED** [review]
Should check interaction with Mateo's changes and fix some probably unintended formatting and debugging changes

File: src/rust_wgpu_backend/code_generator.rs (hunk 3 above)
Comment (Checkmate50) [reply src/rust_wgpu_backend/code_generator.rs:663]: It does, so I put it back

File: src/rust_wgpu_backend/codegen.rs (hunk 2 above)
Comment (Checkmate50) [reply src/rust_wgpu_backend/codegen.rs:541]: oh, uh, yeah, oops.  I'm not actually sure how that got there

File: src/main.rs (hunk 4 above)
Comment (Checkmate50) [reply src/main.rs:49]: makes sense to me

File: src/ir.rs (hunk 5 above)
Comment (Checkmate50) [reply src/ir.rs:373]: I was too scared to change the hashmaps and do the refactoring :/

Part of it is that you would lose the type indication of "FuncletId", which I didn't like, but I can change it

File: src/ir.rs (hunk 6 above)
Comment (Checkmate50) [reply src/ir.rs:203]: Definitely not intentional, that's an IDE thing.  Sorry about that, I'll see if I can fiddle with settings?

File: src/arena.rs (hunk 8 above)
Comment (Checkmate50) [reply src/arena.rs:1]: That's my bad, I did check it all and apparently forgot to delete the actual class in my commit

**Review by od0x0 — APPROVED** [review]

**Pull request merged**